### PR TITLE
[python-package] Allow to pass Arrow table for prediction

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This is the last of a series of PRs towards #6022 and depends on #6167.

For a legible diff, see https://github.com/borchero/LightGBM/compare/arrow-support-init-scores...arrow-support-inference.

### CODE REVIEW

File: tests/python_package_test/test_arrow.py [hunk 1]
```diff
@@ -276,3 +276,17 @@ def test_dataset_construct_init_scores_table():
     actual = dataset.get_init_score()
     expected = init_scores.to_pandas().to_numpy().astype(np.float64)
     np_assert_array_equal(expected, actual, strict=True)
+
+
+# ------------------------------------------ PREDICTION ----------------------------------------- #
+
+
+def test_predict():
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:284]: Can you please add tests for a wider range of `predict()` calls?

https://github.com/microsoft/LightGBM/blob/522f0f07b0eba0e3190c3e5c8e149a205bd20bf3/python-package/lightgbm/basic.py#L1057-L1067

* `num_iteration < {number_of_trees}` = predict with only a subset of trees from the model
* `raw_score=True` = get predictions in units of the objective function, without transformation
* `pred_leaf=True` = for each sample, get the index of the leaf node it falls into in in each tree
* `pred_contrib=True` = SHAP feature contribution values

And for each of those, can you please add tests for each of:

* regression
* binary classification
* multiclass classification
* learning-to-rank

Each combination of those things introduces an opportunity for different types of bugs. I don't think this working only for the default case is sufficient.

Let me know if you need help formulating those test(s).

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py:284]: For an example, look at how the `lightgbm.dask` tests cover these different combinations of `predict()` uses.

https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_dask.py#L287-L298

**Review by jameslamb — CHANGES_REQUESTED** [review]
Changes look good to me! Thanks for the attention to detail you took in following the existing patterns in the project.

But before we merge this, I'd like to see the tests cover more of the `predict()` interface. `predict()` in LightGBM does a lot of different things and its behavior is dependent on the type of machine learning task. I left a suggestion with specific requests for the types of tests I'd like to see.

File: tests/python_package_test/test_arrow.py (hunk 1 above)
Comment (borchero) [reply tests/python_package_test/test_arrow.py]: I extended the tests according to your suggestions :) the test matrix is quite large now but all Arrow tests still only take 14 seconds on my machine.

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py]: > the test matrix is quite large now

Thanks for that! I see some opportunities to cut its size without giving up too much test coverage.

I think trying to force all these combinations into `pytest.parametrize()` on a single test case is leading to the generation of some unnecessary test cases.

For example, `predict()` will only return one prediction type output. I don't think test cases where more than 1 of `raw_score`, `pred_leaf`, and `pred_contrib` are `True` is necessary for this feature. How `lightgbm` handles the precedence of the different prediction types (e.g. how it decides to handle `predict(raw_score=True, pred_leaf=True, pred_contrib=True)` is unrelated to the format of the input data.

Could you please split it up into 4 deep test cases, similar to the examples in that `test_dask.py` file I linked to above?

Like this pseudocode:

```python
def test_predict_regression():
     bst = lgb.train(...)
     p1 = bst.predict(X)
     p1_raw = bst.predict(X, raw_score=True)
     p1_leaf = bst.predict(X, pred_leaf=True)
     p1_pred_contrib = bst.predict(X, pred_contrib=True)
     p1_first_iter = bst.predict(X, start_iteration=0, num_iteration=1, raw_score=True)

     # ... assertions ... #


def test_predict_binary_classification():
    # ...

def test_predict_multiclass_classification():
    # ...

def test_predict_ranking():
    # ...
```

And please use these other approaches to reduce the time it takes the tests to run:

* `{"num_boost_round": 5}` (fewer trees)
* `{"num_leaves": 7}` (shallower trees)

If you'd prefer that I make changes like this and push them to your branch, let me know. I have time this week.

Comment (borchero) [reply tests/python_package_test/test_arrow.py]: All done (I think), tests now only take ~1.5s ;) thanks for the elaborate feedback! 😁

Comment (jameslamb) [reply tests/python_package_test/test_arrow.py]: The new tests look great, thank you for that! Glad they're running fast, and they look like they cover all the combinations I was asking for.

I really like `assert_equal_predict_arrow_pandas()`, thanks for that.

**Review by jameslamb — APPROVED** [review]
Looks great to me, thank you!

@guolinke or @shiyu1994 could you help with one more review on the c_api changes?

**Review by guolinke — APPROVED** [review]
LGTM

**Comment by jameslamb** [comment]
Thank you!

When this builds, I'll merge this. And with that, v4.2.0 will have broad `pyarrow` support 🙌🏻 

Awesome work on this @borchero , thank you so much!!

**Comment by borchero** [comment]
Awesome! @jameslamb: thanks a lot for waiting with v4.2.0 to include Arrow support and, in particular, for all your help getting Arrow support into this package 🫶🏼

**Pull request merged**